File: libspi/image.c
/*
 * AT-SPI - Assistive Technology Service Provider Interface
 * (Gnome Accessibility Project; http://developer.gnome.org/projects/gap)
 *
 * Copyright 2001, 2002 Sun Microsystems Inc.,
 * Copyright 2001, 2002 Ximian, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

/* image.c : implements the Image interface */

#include <config.h>
#include <stdio.h>
#include <libspi/image.h>


SpiImage *
spi_image_interface_new (AtkObject *obj)
{
  SpiImage *new_image = g_object_new (SPI_IMAGE_TYPE, NULL);

  spi_base_construct (SPI_BASE (new_image), G_OBJECT(obj));

  return new_image;
}

static AtkImage *
get_image_from_servant (PortableServer_Servant servant)
{
  SpiBase *object = SPI_BASE (bonobo_object_from_servant (servant));

  g_return_val_if_fail (object, NULL);
  g_return_val_if_fail (ATK_IS_OBJECT(object->gobj), NULL);
  return ATK_IMAGE (object->gobj);
}

static void 
impl_getImagePosition (PortableServer_Servant servant,
		       CORBA_long * x, CORBA_long * y,
		       const CORBA_short coordType,
		       CORBA_Environment *ev)
{
  AtkImage *image = get_image_from_servant (servant);
  gint ix, iy;

  g_return_if_fail (image != NULL);

  atk_image_get_image_position (image,
				&ix, &iy,
				(AtkCoordType) coordType);
  *x = ix;
  *y = iy;
}

static void 
impl_getImageSize (PortableServer_Servant servant,
		   CORBA_long * width, CORBA_long * height,
		   CORBA_Environment *ev)
{
  AtkImage *image = get_image_from_servant (servant);
  gint iw, ih;
  
  g_return_if_fail (image != NULL);

  atk_image_get_image_size (image,
			    &iw, &ih);
  *width = iw;
  *height = ih;
}

static Accessibility_BoundingBox
impl_getImageExtents (PortableServer_Servant servant,
		      const CORBA_short      coordType,
		      CORBA_Environment     *ev)
{
  AtkImage *image;
  gint x, y, width, height;
  Accessibility_BoundingBox bbox;

  bbox.x = bbox.y = bbox.width = bbox.height = -1;

  image = get_image_from_servant (servant);

  if (image)
    {
      atk_image_get_image_size (image, &width, &height);
      atk_image_get_image_position (image, &x, &y, coordType);

      bbox.x = x;
      bbox.y = y;
      bbox.width = width;
      bbox.height = height;
    }

  return bbox;
}

static CORBA_string 
impl__get_imageDescription (PortableServer_Servant servant,
			    CORBA_Environment     *ev)
{
  const char *rv;
  AtkImage   *image = get_image_from_servant (servant);

  g_return_val_if_fail (image != NULL, CORBA_string_dup (""));

  rv = atk_image_get_image_description (image);

  if (rv)
    {
      return CORBA_string_dup (rv);
    }
  else
    {
      return CORBA_string_dup ("");
    }
}

static CORBA_string 
impl__get_imageLocale (PortableServer_Servant servant,
		       CORBA_Environment     *ev)
{
  const char *rv;
  AtkImage   *image = get_image_from_servant (servant);

  g_return_val_if_fail (image != NULL, CORBA_string_dup ("C"));

  rv = atk_image_get_image_locale (image);

  if (rv)
    {
      return CORBA_string_dup (rv);
    }
  else
    {
      return CORBA_string_dup ("C");
    }
}

static void
spi_image_class_init (SpiImageClass *klass)
{
  POA_Accessibility_Image__epv *epv = &klass->epv;

  /* Initialize epv table */
  epv->getImagePosition      = impl_getImagePosition;
  epv->getImageSize          = impl_getImageSize;
  epv->getImageExtents       = impl_getImageExtents;
  epv->_get_imageDescription = impl__get_imageDescription;
  epv->_get_imageLocale      = impl__get_imageLocale;
}

static void
spi_image_init (SpiImage *image)
{
}

BONOBO_TYPE_FUNC_FULL (SpiImage,
		       Accessibility_Image,
		       SPI_TYPE_BASE,
		       spi_image)

File: pyatspi/image.py
#Copyright (C) 2008 Codethink Ltd

#This library is free software; you can redistribute it and/or
#modify it under the terms of the GNU Lesser General Public
#License version 2 as published by the Free Software Foundation.

#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.
#You should have received a copy of the GNU Lesser General Public License
#along with this program; if not, write to the Free Software
#Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

import dbus
from interfaces import *
from accessible import Accessible
from factory import accessible_factory
from accessible import BoundingBox

__all__ = [
           "Image",
          ]

#------------------------------------------------------------------------------

class Image(Accessible):
        """
        An interface implemented by objects which render image data or
        pictorial information to the screen. When onscreen components
        include graphical information that is not purely intended to
        enhance "3d effect" or visual layout, but which conveys some
        semantic or informational content to the sighted user, they should
        implement Image, and that semantic content should be conveyed
        textually to the extent possible via the image description, as
        well as the Accessible::name and Accessible::description properties.
        """

        def getImageExtents(self, coordType):
                """
                Obtain a bounding box which entirely contains the image contents,
                as displayed on screen. The bounds returned do not account for
                any viewport clipping or the fact that the image may be partially
                or wholly obscured by other onscreen content. 
                @param : coordType
                If 0, the returned bounding box position is returned relative
                to the screen; if 1, the bounding box position is returned relative
                to the containing window. 
                @return a BoundingBox enclosing the image's onscreen representation.
                """
                func = self.get_dbus_method("getImageExtents", dbus_interface=ATSPI_IMAGE)
                return BoundingBox(*func(dbus.Int16(coordType)))

        def getImagePosition(self, coord_type):
                """
                Get the coordinates of the current image position on screen.
                @param : x
                Back-filled with the x coordinate of the onscreen image (i.e.
                the minimum x coordinate) 
                @param : y
                Back-filled with the y coordinate of the onscreen image (i.e.
                the minimum y coordinate) 
                @param : coordType
                If 0, the returned x and y coordinates are returned relative
                to the screen; if 1, they are returned relative to the containing
                window.
                """
                func = self.get_dbus_method("getImagePosition", dbus_interface=ATSPI_IMAGE)
                return func(coord_type)

        def getImageSize(self):
                """
                Obtain the width and height of the current onscreen view of the
                image. The extents returned do not account for any viewport clipping
                or the fact that the image may be partially or wholly obscured
                by other onscreen content. 
                @param : width
                Back-filled with the x extents of the onscreen image (i.e. the
                image width in pixels) 
                @param : height
                Back-filled with the y extents of the onscreen image (i.e. the
                image height in pixels)
                """
                func = self.get_dbus_method("getImageSize", dbus_interface=ATSPI_IMAGE)
                return func()

        def get_imageDescription(self):
                return dbus.String(self._pgetter(self._dbus_interface, "imageDescription"))
        _imageDescriptionDoc = \
                """
                A UTF-8 string providing a textual description of what is visually
                depicted in the image.
                """
        imageDescription = property(fget=get_imageDescription, doc=_imageDescriptionDoc)

        def get_imageLocale(self):
                return dbus.String(self._pgetter(self._dbus_interface, "imageLocale"))
        _imageLocaleDoc = \
                """
                A string corresponding to the POSIX LC_MESSAGES locale used by
                the imageDescription.
                """
        imageLocale = property(fget=get_imageLocale, doc=_imageLocaleDoc)

# Register the accessible class with the factory.
accessible_factory.register_accessible_class(ATSPI_IMAGE, Image)

#END----------------------------------------------------------------------------
